sycl/source/detail/scheduler/commands.cpp

```diff
@@ -513,13 +512,7 @@ Command *Command::processDepEvent(EventImplPtr DepEvent, const DepDesc &Dep) {
       getType() != CommandType::HOST_TASK)
     return nullptr;
 
-  ContextImplPtr DepEventContext = DepEvent->getContextImpl();
-  // If contexts don't match we'll connect them using host task
-  if (DepEventContext != WorkerContext && !WorkerContext->is_host()) {
-    Scheduler::GraphBuilder &GB = Scheduler::getInstance().MGraphBuilder;
-    ConnectionCmd = GB.connectDepEvent(this, DepEvent, Dep);
-  } else
-    MPreparedDepsEvents.push_back(std::move(DepEvent));
+  MPreparedDepsEvents.push_back(std::move(DepEvent));
 
   return ConnectionCmd;
```

Comment:
Sorry, it didn't click in my head the last time I reviewed this. `ConnectionCmd` is always `nullptr` now. Is there any point in returning it at all now?

@romanovvlad

Reply:
I wanted to check with you whether ConnectionCmd here has any use cases that I am not aware of before making further changes in this WIP PR.  Making ConnectionCmd always nullptr doesn't break any tests using the cuda backend.  If ConnectionCmd has no purpose then I can change the function to return void but this requires quite a few cascading changes to other functions that would then also return void.
It would appear that it is the responsibility of graph_builder to insure that memory is located in the correct context.  It then doesn't appear appropriate that this command class method also takes on this responsibility as a kind of "redundancy".  If ConnectionCmd has no usage then I think that it would be good to remove it since, for example, it introduced unexpected/unwanted effects when I made the changes in graph_builder.cpp introducing Peer to Peer copy.

Reply:
@alexbatashev

I've investigated what happens when I also make `processDepEvent` and `addDep` return `void` (no such changes committed), which as you write should make no difference since `ConnectionCmd` is always `nullptr` in this PR.  Such changes remove a lot of code from graph_builder.cpp, commands.cpp and commands.hpp.  Interestingly, making these changes leads to many tests failing in the test-suite when no tests failed with processDepEvent returning nullptr.

I do not know whether the test coverage is simply insufficient (presumably there should be cases checking that `ConnectionCmd` works as expected and is not nullptr if there are use cases) or whether there really is no use case for `ConnectionCmd` for all possible graphs that can be created.  If `ConnectionCmd` really is redundant it would be useful to remove it since this would make it possible to simplify command and graph_builder considerably.  If `ConnectionCmd` is not redundant then I do not think that the cross context optimization introduced in this PR is going to be work without considerably more changes than currently included in this PR (unless `ConnectionCmd` is redundant only for the cuda backend).  

Currently the optimizations do not do anything anyway due to #4571, but #4571 is a temporary fix and is expected to be reverted eventually.